common/rpc/interceptor/telemetry.go

```diff
@@ -246,23 +249,50 @@ func (ti *TelemetryInterceptor) emitActionMetric(
 			return
 		}
 
+		messagesById := collection.NewIndexedTakeList(
+			completedRequest.Messages,
+			func(msg *protocolpb.Message) string { return msg.Id })
+
+		var updates int64
 		hasMarker := false
 		for _, command := range completedRequest.Commands {
-			if _, ok := commandActions[command.CommandType]; ok {
-				switch command.CommandType {
-				case enums.COMMAND_TYPE_RECORD_MARKER:
-					// handle RecordMarker command, they are used for localActivity, sideEffect, versioning etc.
-					hasMarker = true
-				case enums.COMMAND_TYPE_START_CHILD_WORKFLOW_EXECUTION:
-					// Each child workflow counts as 2 actions. We use separate tags to track them separately.
-					metrics.ActionCounter.With(metricsHandler).Record(1, metrics.ActionType("command_"+command.CommandType.String()))
-					metrics.ActionCounter.With(metricsHandler).Record(1, metrics.ActionType("command_"+command.CommandType.String()+"_Extra"))
-				default:
-					// handle all other command action
-					metrics.ActionCounter.With(metricsHandler).Record(1, metrics.ActionType("command_"+command.CommandType.String()))
+			if _, ok := commandActions[command.CommandType]; !ok {
+				continue
+			}
+
+			switch command.CommandType { // nolint:exhaustive
+			case enums.COMMAND_TYPE_RECORD_MARKER:
+				// handle RecordMarker command, they are used for localActivity, sideEffect, versioning etc.
+				hasMarker = true
+			case enums.COMMAND_TYPE_START_CHILD_WORKFLOW_EXECUTION:
+				// Each child workflow counts as 2 actions. We use separate tags to track them separately.
+				metrics.ActionCounter.With(metricsHandler).Record(1, metrics.ActionType("command_"+command.CommandType.String()))
+				metrics.ActionCounter.With(metricsHandler).Record(1, metrics.ActionType("command_"+command.CommandType.String()+"_Extra"))
+			case enums.COMMAND_TYPE_PROTOCOL_MESSAGE:
+				msg, _ := messagesById.Take(command.GetProtocolMessageCommandAttributes().GetMessageId())
+				if msg == nil || msg.Body == nil {
+					continue
+				}
+				body, err := msg.Body.UnmarshalNew()
+				if err != nil {
+					continue
 				}
+				switch body.(type) {
```

Comment:
Doesn't `updatepb.Outcome` also be counted as an update operation?

Reply:
nvm. Just read your test cases. You are explicitly not counting `Outcome` types.